source/keyring_trace.c

```diff
@@ -15,11 +15,43 @@
 #include <aws/cryptosdk/private/keyring_trace.h>
 #include <aws/cryptosdk/private/utils.h>
 
+bool aws_cryptosdk_keyring_trace_is_valid(struct aws_array_list *trace) {
+    if (!trace) {
+        return false;
+    }
+    /* iterate over each record in the list */
+    size_t num_records = aws_array_list_length(trace);
+    bool is_valid      = aws_array_list_is_valid(trace);
+    for (size_t idx = 0; idx < num_records; ++idx) {
```

Comment:
Having a loop in an `is_valid()` function is unexpected, and should be called out in the documentation. It may make it too expensive for use in preconditions.